Problem finding nc_def_var_szip for --disable-shared builds linking against static only netcdf-c library

An email from George at NOAA installing on NOAA Cray:


```
When I build netcdfp/4.7.4 on most systems the include file mods are
cd lib/netcdfp
SURGE-slogin1 > ls -l include/*.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel 269722 May  5 17:15 include/netcdf.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel    642 May  5 17:15 include/netcdf4_f03.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel 239096 May  5 17:15 include/netcdf4_nc_interfaces.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel  42133 May  5 17:15 include/netcdf4_nf_interfaces.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel    538 May  5 17:15 include/netcdf_f03.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel 222833 May  5 17:15 include/netcdf_fortv2_c_interfaces.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel  28228 May  5 17:15 include/netcdf_nc_data.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel 177111 May  5 17:15 include/netcdf_nc_interfaces.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel  58363 May  5 17:15 include/netcdf_nf_data.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel 104236 May  5 17:15 include/netcdf_nf_interfaces.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel   1382 May  5 17:15 include/typesizes.mod


On surge  I get
../netcdfp/include/*mod
-rw-r--r-- 1 George.Vandenberghe emcmodel 218741 May  5 18:46 ../netcdfp/include/netcdf.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel    538 May  5 18:46 ../netcdfp/include/netcdf_f03.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel  18282 May  5 18:46 ../netcdfp/include/netcdf_nc_data.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel 177111 May  5 18:46 ../netcdfp/include/netcdf_nc_interfaces.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel  36526 May  5 18:46 ../netcdfp/include/netcdf_nf_data.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel 104236 May  5 18:46 ../netcdfp/include/netcdf_nf_interfaces.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel   1382 May  5 18:46 ../netcdfp/include/typesizes.mod

even with an apparently clean build.

This also happens on another system when I try to disable shared in the configure.   On the crays shared HAS to be disabled
or nothing builds

The missing mods are 

-rw-r--r-- 1 George.Vandenberghe emcmodel    642 May  5 17:15 include/netcdf4_f03.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel 239096 May  5 17:15 include/netcdf4_nc_interfaces.mod
-rw-r--r-- 1 George.Vandenberghe emcmodel  42133 May  5 17:15 include/netcdf4_nf_interfaces.mod

-rw-r--r-- 1 George.Vandenberghe emcmodel 222833 May  5 17:15 include/netcdf_fortv2_c_interfaces.mod
```

The challenge is, I'm not really sure what the right thing to do here is.

The problem is that the netcdf-fortran configure is not finding all the dependent libraries of netcdf-c when attempting to link a sample program. So the test program in config.log is failing like this:

```
configure:22668: gcc -o conftest -g -O2 -I/usr/local/netcdf-c-4.7.4_hdf5_1.10.6_no_shared/include -L/usr/local/netcdf-c-4.7.4_hdf5_1.10.6_no_shared/lib conftest.c -lnetcdf  -ldl -lm  >&5
/usr/local/netcdf-c-4.7.4_hdf5_1.10.6_no_shared/lib/libnetcdf.a(libnchdf5_la-hdf5file.o): In function `sync_netcdf4_file':
/home/ed/Downloads/netcdf-c-4.7.4/libhdf5/hdf5file.c:222: undefined reference to `H5Fflush'
/usr/local/netcdf-c-4.7.4_hdf5_1.10.6_no_shared/lib/libnetcdf.a(libnchdf5_la-hdf5file.o): In function `nc4_close_netcdf4_file':
/home/ed/Downloads/netcdf-c-4.7.4/libhdf5/hdf5file.c:273: undefined reference to `H5Fclose'
/usr/local/netcdf-c-4.7.4_hdf5_1.10.6_no_shared/lib/libnetcdf.a(libnchdf5_la-hdf5file.o): In function `dumpopenobjects':
/home/ed/Downloads/netcdf-c-4.7.4/libhdf5/hdf5file.c:377: undefined reference to `H5Fget_obj_count'
/usr/local/netcdf-c-4.7.4_hdf5_1.10.6_no_shared/lib/libnetcdf.a(libnchdf5_la-hdf5filter.o): In function `nc4_global_filter_action':
/home/ed/Downloads/netcdf-c-4.7.4/libhdf5/hdf5filter.c:134: undefined reference to `H5Zregister'
/home/ed/Downloads/netcdf-c-4.7.4/libhdf5/hdf5filter.c:149: undefined reference to `H5Zunregister'

```

Now one answer is to tell the user to set LIBS=`nc-config --libs`

Is that what we want to do?

Or do we want netcdf-fortran configure to run nc-config and get the answer for the user?

I'm not sure what the autoconf custom would be here.@skosukhin I would welcome your advice here...

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -14,8 +14,8 @@ Entries are in reverse chronological order (most recent first).
 
 ### Changes
 
-* Added support for HDF5 compact storage. See [Github #213](https://github.com/Unidata/netcdf-fortran/issues/207).
-* Added support for creating netCDF/HDF5 files with szip compression with new functions nf90_def_var_szip() and nf_def_var_szip(). See [Github #213](https://github.com/Unidata/netcdf-fortran/issues/213).
+* Added support for HDF5 compact storage in F77 API. See [Github #213](https://github.com/Unidata/netcdf-fortran/issues/207).
+* Added support for creating netCDF/HDF5 files with szip compression with new functions nf90_def_var_szip() and nf_def_var_szip(). See [Github #213](https://github.com/Unidata/netcdf-fortran/issues/213), [Github #251](https://github.com/Unidata/netcdf-fortran/issues/251).
 * Corrected an issue where parallel netCDF-Fortran builds would fail despite the presense of MPI libraries/compiler/infrastructure. See [Github #208](https://github.com/Unidata/netcdf-fortran/issues/208) for more information.
 
 ## 4.5.2 - September 18, 2019

--- a/configure.ac
+++ b/configure.ac
@@ -89,8 +89,6 @@ if test "x$nc_has_parallel4" = xyes; then
 fi
 AC_MSG_RESULT([$nc_has_parallel4])
 
-# Starting with version 4.7.4, netcdf-c has function nc_def_var_szip().
-AC_SEARCH_LIBS([nc_def_var_szip], [netcdf], [], [AC_MSG_ERROR([netcdf-c version 4.7.4 or greater is required])])
 AC_MSG_CHECKING([if netCDF was built with support for szip writes])
 AC_COMPILE_IFELSE([AC_LANG_PROGRAM([], [[
 #include <netcdf_meta.h>
@@ -402,10 +400,16 @@ AC_SEARCH_LIBS([dlopen], [dl dld], [], [])
 
 # Find the netCDF header and library.
 AC_CHECK_HEADERS([netcdf.h], [], [AC_MSG_ERROR([netcdf.h could not be found. Please set CPPFLAGS.])])
-AC_SEARCH_LIBS([nc_open], [netcdf], [], [AC_MSG_ERROR([Could not link to netcdf C library. Please set LDFLAGS.])])
+AC_SEARCH_LIBS([nc_open], [netcdf], [],
+			  [AC_MSG_ERROR([Could not link to netcdf C library. Please set LDFLAGS; for static builds set LIBS to the results of nc-config --libs.])])
 
 # See if various functions are available
-AC_CHECK_FUNCS([nc_def_opaque nccreate nc_set_log_level oc_open])
+AC_CHECK_FUNCS([nc_def_opaque nccreate nc_set_log_level oc_open nc_def_var_szip])
+
+# Starting with version 4.7.4, netcdf-c has function nc_def_var_szip().
+if test "x$ac_cv_func_nc_def_var_szip" != xyes; then
+   AC_MSG_ERROR([netcdf-c version 4.7.4 or greater is required])
+fi
 
 nc_build_v2=no
 nc_build_v4=no

